### PR TITLE
feat(elements-core): make sidebar resizable

### PR DESCRIPTION
**Pull request by porotkin**
# Feature: Resizable Sidebar

I consider that it'd be better if a user could resize the sidebar to fit his needs.
I need this functionality myself, so I'd be glad if you take a look at my PR, @Amjcraft

Thank you! 

In general, make sure you have: (check the boxes to acknowledge you've followed this template)

- [x] Read [`CONTRIBUTING.md`](../CONTRIBUTING.md)

#### Other Available PR Templates:

- Release: https://github.com/stoplightio/elements/compare?template=release.md
  - [x] [Read the release section of `CONTRIBUTING.md`](../CONTRIBUTING.md#releasing-elements)

### CODE REVIEW

File: packages/elements-core/src/hooks/useResizer.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import React, { useEffect, useRef, useState } from 'react';
+
+export function useResizer(sidebarWidth: number): [React.Ref<HTMLDivElement>, number, (event: any) => void] {
```
Comment (aerialist7) [inline packages/elements-core/src/hooks/useResizer.ts:3]: Seems this hook consider only to `SidebarLayout` component and should be private

Comment (aerialist7) [inline packages/elements-core/src/hooks/useResizer.ts:3]: What about type aliases?

File: packages/elements-core/src/hooks/useResizer.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  const sidebarRef = useRef<HTMLDivElement | null>(null);
```
Comment (aerialist7) [inline packages/elements-core/src/hooks/useResizer.ts:4]: You are using `React.` prefix somewhere.
But here - don't.

Seems you need to align the approach

**Review by aerialist7 — CHANGES_REQUESTED** [review]

**Comment by porotkin** [comment]
@P0lip, please, take a loot at the changes. I have accidentally removed the request for you

**Review by aerialist7 — APPROVED** [review]

**Comment by P0lip** [comment]
Thanks for the PR! I'll take a look at it shortly.

File: packages/elements-core/src/core.css [hunk 3]
```diff
@@ -85,3 +85,15 @@
...
+
+.sl-sidebar-resizer {
+  cursor: col-resize;
+  width: 1px;
+  flex-grow: 0;
+  flex-shrink: 0;
+  flex-basis: 6px;
+  justify-self: flex-end;
+}
+
```
Comment (P0lip) [inline packages/elements-core/src/core.css:99]:
```suggestion

```

File: packages/elements-core/src/components/Layout/SidebarLayout.tsx [hunk 4]
```diff
@@ -25,27 +27,74 @@ export const SidebarLayout = React.forwardRef<HTMLDivElement, SidebarLayoutProps
...
-          pinY
-          overflowY="auto"
-          style={{
-            width: `calc((100% - ${maxContentWidth}px) / 2 + ${sidebarWidth}px)`,
-            paddingLeft: `calc((100% - ${maxContentWidth}px) / 2)`,
-            minWidth: `${sidebarWidth}px`,
-          }}
+          ref={sidebarRef}
+          direction="row"
+          onMouseDown={(e: { preventDefault: () => void }) => e.preventDefault()}
```
Comment (P0lip) [inline packages/elements-core/src/components/Layout/SidebarLayout.tsx:31]:
```suggestion
          onMouseDown={(e: React.MouseEvent<HTMLElement>) => e.preventDefault()}
```

File: packages/elements-core/src/components/Layout/SidebarLayout.tsx [hunk 5]
```diff
@@ -25,27 +27,74 @@ export const SidebarLayout = React.forwardRef<HTMLDivElement, SidebarLayoutProps
...
-          pos="sticky"
-          pinY
-          overflowY="auto"
-          style={{
-            width: `calc((100% - ${maxContentWidth}px) / 2 + ${sidebarWidth}px)`,
-            paddingLeft: `calc((100% - ${maxContentWidth}px) / 2)`,
-            minWidth: `${sidebarWidth}px`,
-          }}
+          ref={sidebarRef}
+          direction="row"
```
Comment (P0lip) [inline packages/elements-core/src/components/Layout/SidebarLayout.tsx:30]:
```suggestion
```
that's already a default, isn't it?

File: packages/elements-core/src/components/Layout/SidebarLayout.tsx [hunk 6]
```diff
@@ -25,27 +27,74 @@ export const SidebarLayout = React.forwardRef<HTMLDivElement, SidebarLayoutProps
...
+      if (isResizing) {
+        setCurrentSidebarWidth(mouseMoveEvent.clientX - sidebarRef.current!.getBoundingClientRect().left);
+      }
+    },
+    [isResizing],
+  );
+
+  React.useEffect(() => {
+    window.addEventListener('mousemove', resize);
+    window.addEventListener('mouseup', stopResizing);
```
Comment (P0lip) [inline packages/elements-core/src/components/Layout/SidebarLayout.tsx:91]:
```suggestion
    window.addEventListener('mouseup', stopResizing, { passive: true });
```
could be passive

**Review by P0lip — COMMENTED** [review]
Looks good overall. Nice work. Thank you!

**Comment by porotkin** [comment]
Thanks for the tips and suggestions! I've merged 'em all. It seems that the CI is failing but the reason is not related to the changes (it cannot download tslib somehow)

**Review by aerialist7 — APPROVED** [review]

**Review by P0lip — APPROVED** [review]
Thank you!

**Comment by porotkin** [comment]
@P0lip, thanks for the guidance :)

By the way, could you, please, take a glance on https://github.com/stoplightio/elements/issues/2253? We'd really appreciate any recommendation/workaround/tip regarding this matter.

And currently I am looking into AsyncApi rendering support in @stoplight/elements. I wrote a comment https://github.com/stoplightio/elements/issues/1766#issuecomment-1290277109. If it's of interest to you, I'd be glad to receive any feedback. Currently, I've made some progress transforming AsyncApi specification into ServiceNode and rendered Streetlight Kafka API:
image

**Comment by P0lip** [comment]
#2253 is probably due to us trying to create a regular expression out of it.
We use `.match` somewhere and that could explain. I believe one can just use `includes` in there.

By the way, could you make the gutter a bit wider? It's a bit difficult to hit right now.

**Comment by mnaumanali94** [comment]
Thanks @porotkin for this! In addition to what Jakub said, i think the resizing could use a smaller limit in order to not let people mess up the layout

**Comment by porotkin** [comment]
Thanks, @P0lip and @mnaumanali94, I have enlarged the resizer to 1em instead of 1px and set the bounds for currentSidebarWidth property. Hopefully, this is what you wanted to see.

**Comment by P0lip** [comment]
Thank you!

**Pull request merged**